gpt_index/readers/file/markdown_parser.py

```diff
@@ -0,0 +1,90 @@
+
+
+from abc import abstractmethod
+from pathlib import Path
+from typing import Dict, List, Optional, Union
+import re
+
+class MarkdownParser:
+    """Base class for all parsers."""
+
+    def __init__(self, parser_config: Optional[Dict] = {'remove_hyperlinks': True, 'remove_images': True}):
+        """Init params."""
+        self._parser_config = parser_config
+
+    def init_parser(self) -> Dict:
+        """Init parser and store it."""
+        parser_config = self._init_parser()
+        self._parser_config = parser_config
+
+    def markdown_to_dict(self, markdown_text):
+      """Convert a markdown file to a dictionary. The keys are the headers and the values are the text under each header."""
+      markdown_dict = {}
+      lines = markdown_text.split('\n')
+
+      current_header = None
+      current_text = ''
+
+      for line in lines:
+        header_match = re.match(r'^#+\s', line)
+        if header_match:
+          if current_header is not None:
+            if current_text == '' or None: continue
+            markdown_dict[current_header] = current_text
+          
+          current_header = line
+          current_text = ''
+        else: current_text += line + '\n'
+      markdown_dict[current_header] = current_text
+
+      if current_header is not None:
+        markdown_dict = {re.sub(r'#', '', key).strip(): re.sub(r'<.*?>', '', value) for key, value in markdown_dict.items()}
+      else:
+        markdown_dict = {key: re.sub('\n', '', value) for key, value in markdown_dict.items()}
+
+      return markdown_dict
+    
+    def remove_images(self, content):
+      """Get a dictionary of a markdown file from its path"""
+      pattern = r"!{1}\[\[(.*)\]\]"
+      content = re.sub(pattern, "", content)
+      return content
+
+    def remove_hyperlinks(self, content):
+      """Get a dictionary of a markdown file from its path"""
+      pattern = r"\[(.*?)\]\((.*?)\)"
+      content = re.sub(pattern, r"\1", content)
+      return content
+
+    @property
+    def parser_config_set(self) -> bool:
+        """Check if parser config is set."""
+        return self._parser_config is not None
+
+    @property
+    def parser_config(self) -> Dict:
+        """Check if parser config is set."""
+        if self._parser_config is None:
+            raise ValueError("Parser config not set.")
+        return self._parser_config
+
+    @abstractmethod
+    def _init_parser(self) -> Dict:
+        """Initialize the parser with the config."""
+        parser_config = {'remove_hyperlinks': True, 'remove_images': True}
```

Comment:
note: you don't need to put this in _init_parser (this is really meant for loading "big" stuff like downloading a model from huggingface), you can just put these params as part of the `__init__`. Here you can just return an empty dictionary